tests/tool_excimer_sample_set_test.php

```diff
@@ -0,0 +1,183 @@
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class tool_excimer_sample_set_test extends excimer_testcase {
+
+    /**
+     * Tests adding samples to the object.
+     */
+    public function test_add_sample() {
+        $samples = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+        ];
+
+        $set = new sample_set('a', 0, 1024);
+
+        foreach ($samples as $sample) {
+            $set->add_sample($sample);
+        }
+
+        $this->assertEquals($samples, $set->samples);
+    }
+
+    /**
+     * Tests the effect of filtering while adding samples.
+     */
+    public function test_filtering() {
+        $samples = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+        ];
+        // This is every 2nd element of $samples.
+        $expected1 = [
+            $samples[1],
+            $samples[3]
+        ];
+        // This is every 4th element of $samples.
+        $expected2 = [
+            $samples[3]
+        ];
+
+        $set = new sample_set('a', 0, 1024);
+
+        // Each time this is called, the filter rate is doubled.
+        $set->apply_doubling();
+
+        // Filter rate should be 2, thus, only every 2nd sample should be recorded in sample set.
+
+        foreach ($samples as $sample) {
+            $set->add_sample($sample);
+        }
+
+        // Only every 2nd sample should be recorded in sample set.
+        $this->assertEquals($expected1, $set->samples);
+
+        $set = new sample_set('a', 0, 1024);
+        $set->apply_doubling();
+        $set->apply_doubling();
+
+        // Filter rate should be 4, thus, only every 4th sample should be recorded in sample set.
+
+        foreach ($samples as $sample) {
+            $set->add_sample($sample);
+        }
+
+        $this->assertEquals($expected2, $set->samples);
+    }
+
+    /**
+     * Tests stripping existing samples when calling apply_doubling.
+     */
+    public function test_stripping() {
+        $samples = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+        ];
+        // This is $samples ofter being stripped once.
+        $expected1 = [
+            $samples[1],
+            $samples[3]
+        ];
+        // This is $samples ofter being stripped twice.
+        $expected2 = [
+            $samples[3]
+        ];
+
+        $set = new sample_set('a', 0, 1024);
+
+        foreach ($samples as $sample) {
+            $set->add_sample($sample);
+        }
+
+        $set->apply_doubling();
+
+        // Half of the samples should be stripped (every 2nd).
+        $this->assertEquals($expected1, $set->samples);
```

Comment:
Maybe this format would read a bit better? (What you intend to do and what the expected behaviours are)
```suggestion
        // Every 2nd sample should be stripped after doubling.
        $set->apply_doubling();
        $this->assertEquals($expected1, $set->samples);
```